[Compose 페이먼츠 3단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 오이 ! 3단계도 잘 부탁드립니다 👍 

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

https://github.com/user-attachments/assets/8849fe1d-bb92-47b6-a23f-574d6941d10d

chanho0908 committed:
docs: 3단계 기능 요구 사항 추가

chanho0908 committed:
feat: 카드사 로고 이미지 추가

chanho0908 committed:
feat: 카드사별 색상 추가

chanho0908 committed:
feat: 카드사 문자열 리소스 추가

chanho0908 committed:
feat: 카드사 종류를 나타내는 Banks Enum 클래스 추가

chanho0908 committed:
feat: 카드 정보 도메인 모델에 은행 정보 추가

chanho0908 committed:
feat: 카드사 아이콘 컴포넌트 구현

chanho0908 committed:
feat: 은행 선택 BottomSheet 구현

- 은행 선택 BottomSheet UI를 구현합니다.
- 모든 은행 아이콘과 이름이 표시되는지 확인하는 테스트를 추가합니다.

chanho0908 committed:
feat: 회사 리소스 프로바이더 구현

카드 회사별 이름, 서명 색상, 아이콘을 제공하는 `CompanyResourceProvider` 클래스를 추가합니다.

chanho0908 committed:
feat: 카드사명 컴포넌트 구현

chanho0908 committed:
refactor: CardType을 sealed interface로 변경하고 CompanyResourceProvider 추가

chanho0908 committed:
feat: 카드 은행 선택 BottomSheet 구현

- 카드 은행 선택 BottomSheet를 구현합니다.
- 카드 은행 선택에 따른 카드 미리보기 UI를 변경합니다.
- 카드 정보 상태 저장 로직을 개선합니다.

chanho0908 committed:
feat: 카드사에 따라 다른 카드 이름 표시 기능 추가

- RegisteredCard 컴포넌트에 카드사 이름을 표시합니다.
- BankType을 Card 도메인에 추가합니다.
- CompanyResourceProvider를 사용하여 카드사 이름을 가져옵니다.

chanho0908 committed:
feat: 은행 및 카드 UI 상태 정의

chanho0908 committed:
feat: 카드 등록 화면에서 은행 선택 BottomSheet 닫기 기능 추가

chanho0908 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -1,14 +1,37 @@
     val ownerName: String,
     val password: String,
+    val bank: BankState,
 ) {
```
오마이갓 이건 수정할게요 52

chanho0908 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -1,14 +1,37 @@
 package woowacourse.payments.domain
 
+import woowacourse.payments.ui.state.BankState
+
```
오이 여기 Domain이 ui에 의존하고 있는 문제는 인지하고 수정하고 있으니 리뷰에서 제외해주셔도 될 것 같아요 !

chanho0908 committed:
refactor: `Banks` enum을 `CardCompany`로 이름 변경

chanho0908 committed:
refactor: BankState 네이밍 CardCompanyState로 변경

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/cards/CardsScreenTest.kt:
```diff
@@ -50,12 +54,18 @@ class CardsScreenTest {
+                uiState,
+                Event(CardScreenUiEvent.Idle),
+                {},
+            )
```
왜 `resourceProvider`만 차별하나요..

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/component/NewCardNameTest.kt:
```diff
@@ -0,0 +1,28 @@
+import woowacourse.payments.ui.component.NewCardName
+
+class NewCardNameTest {
+    @get:Rule
+    val composeTestRule = createAndroidComposeRule<ComponentActivity>()
```
일반적인 `createComposeRule()`과의 차이점이 무엇인가요??

chanho0908 replied:
`createComposeRule()` 는 순수 `Compose` 환경을 위한 `Rule`이에요 !

때문에 안드로이드 프레임워크 의존성이 없어  `Activity`나 `Context`, 안드로이드 리소스에 접근할 수 없습니다

`createAndroidComposeRule` Android Ui Test에서 사용하던 `ActivityScenarioRule`이 포함되어 있어
Android Resource에 접근이 가능하다고 하네요 :)

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/new/NewCardUiStateHolderTest.kt:
```diff
@@ -74,18 +72,16 @@ class NewCardUiStateHolderTest {
+                },
+            ) {
+                saver.run { save(holder) }
+            }
```
Saver 테스트 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/component/CardCompanyIcon.kt:
```diff
@@ -0,0 +1,71 @@
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun CardCompanyIcon(
+    bankIcon: Int,
```
`bankIcon`이 리소스 ID라는 것을 명시하면 좋을 것 같아요!

chanho0908 replied:
`CardComapny`를 직접 주입받아 확장함수로 매핑하도록 수정했어요 !

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -6,26 +6,47 @@ import androidx.compose.foundation.layout.Box
+                    color = backgroundColor,
                     shape = RoundedCornerShape(5.dp),
-                ).clickable(onClick = { onClick(cardType) }),
+                ).clickable(onClick = { onClick(card) }),
```
카드의 클릭 여부를 `PaymentCard` 내부에서 결정하게끔 코드를 작성한 페토만의 이유가 있나요?

chanho0908 replied:
https://github.com/woowacourse/android-payments/pull/68/commits/26b832d99598bf5fb7e93409e693fefb29a30c2f

요것두 제가 놓친 부분이네요 해당 컴포저블이 재사용되는 만큼 클릭 여부는 상위에서 결정하면 훨씬 사용처에 맞게 분리될 것 같습니다 👍

chanho0908 committed:
refactor: `CardCompanyState`를 `CardCompany`로 변경

chanho0908 committed:
refactor: `CompanyResourceProvider`에서 카드 회사 이름 반환 시 null을 허용하지 않도록 수정

chanho0908 committed:
refactor: NewCardUiStateHolder에서 Card 객체 직접 사용하지 않도록 수정

chanho0908 committed:
refactor: NewCardUiState에서 Card 객체 분리

chanho0908 committed:
refactor: CardState.Registered 사용하도록 CardsScreen UI 로직 수정

chanho0908 committed:
refactor: `CardState.Registered`가 `Card` 도메인 모델을 직접 사용하도록 변경

chanho0908 committed:
refactor: `NewCardUiState`에 카드 상태 및 변환 로직 추가

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/new/BankSelectBottomSheetTest.kt:
```diff
@@ -0,0 +1,49 @@
+                    bankName,
+                ).assertIsDisplayed()
+        }
+    }
```
`BankSelectRow`가 은행 목록을 인자로 받게 하면, 테스트가 더 수월하지 않을까요?
반영하지는 않으셔도 돼요 ㅎㅎ

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/core/CompanyResourceProvider.kt:
```diff
@@ -0,0 +1,51 @@
+            CardCompany.KB -> SignatureKbBrown
+        }
+
+    fun getCompanyIcon(bank: CardCompany): Int =
```
다 같은 타입인데 매개변수의 이름을 각각 다르게 하신 이유가 있을까요?

chanho0908 replied:
bank -> company로 네이밍을 변경하는 과정에서 누락된 것 같아유 😢

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/component/RegisteredCard.kt:
```diff
@@ -26,12 +28,17 @@ fun RegisteredCard(
     expireDateSeparator: String,
     modifier: Modifier = Modifier,
 ) {
+    val resourceProvider = CompanyResourceProvider()
```
`CompanyResourceProvider`의 인스턴스가 여기저기서 불필요하게 생성되고, 또 리컴포지션이 발생하면 재생성되는 것 같은데요.
차라리 싱글톤으로 사용하는 것은 어떤가요?

chanho0908 replied:
좋은 의견 고마워요 🥒 

매번 CompanyResourceProvider가 필요한 컴포저블마다 객체를 생성하고 이를 관리하는 과정이
너무 복잡하게 느껴져서 확장함수로 수정했어요 !

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/BankSelectBottomSheet.kt:
```diff
@@ -0,0 +1,104 @@
+                onClick = { bankType ->
+                    selectedBank = bankType
+                    onBankSelect(bankType)
+                },
```
저는 `LaunchedEffect` 내부에서 `hide()`를 호출함과 동시에 `onBankSelect`를 호출했는데요.
페토는 `BankSelectRow`의 클릭 동작에 `onBankSelect`까지 호출한 이유가 있으신가요?

chanho0908 replied:
[a094801](https://github.com/woowacourse/android-payments/pull/68/commits/a094801c0b03b06751a2f4b850b313c841de93f1)

이 부분은 제가 `LaunchedEffect`와 `BottomSheet`의 상태 관리에 대한 이해가 부족했던 것 같아요

수정된 코드를 보면 `BottomSheet`의 상위 컴포저블의 `LaunchedEffect`에서 `CardCompnay`의 상태를 관리하도록 수정했습니다 

이렇게 되니, `BottomSheet`는 화면을 렌더링하는데만 집중하게 되어 `Stateless`한 객체가 되었군요 !
🥒 덕에 좋은 관점을 배워갑니다 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardScreen.kt:
```diff
@@ -58,52 +66,69 @@ private const val CARD_EXPIRE_DATE_SEPARATOR = " / "
+    val companyName: String =
+        company?.let { company ->
+            stringResource(resourceProvider.getCompanyName(company))
+        } ?: ""
```
카드사 이름이 존재하지 않으면 `NewCardName`을 UI 트리에 추가하지 않기 
vs
카드사 이름이 존재하지 않으면 빈 문자열을 출력하기

chanho0908 replied:
[70997b0](https://github.com/woowacourse/android-payments/pull/68/commits/70997b05d756e1910692715e81c7c8dc6676b96f)

👀 오!! 이건 제가 완전 놓친 부분이었네요 좋은 지적 감사합니다 🥒 
카드사 이름이 존재할때만 렌더링 되도록 수정했습니다 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardUiState.kt:
```diff
@@ -1,7 +1,44 @@
+                    )
+            }
+
+    fun toDomain(): Card =
+        Card(
+            number = number,
+            expireDate = expireDate,
+            ownerName = ownerName,
+            password = password,
+            company =
+                (company as? CardCompanyState.Selected)?.company
+                    ?: throw IllegalStateException(CARD_COMPANY_NOT_SELECTED),
+        )
```
카드사가 선택되지 않았을 때 추가 버튼을 누르면 (그럴 일은 거의 없겠지만..) 예외가 발생하는데,
이 경우 외부에서 따로 처리하지 않고 그냥 앱을 터뜨리고 계시네요?
더 안전한 방법은 없을까요?

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardUiStateHolder.kt:
```diff
@@ -17,30 +18,69 @@ class NewCardUiStateHolder(
+        private const val KEY_PASSWORD = "password"
+        private const val KEY_COMPANY = "company"
+
+        val Saver: Saver<NewCardUiStateHolder, Any> =
+            mapSaver(
```
`mapSaver`를 사용하신 이유는 무엇인가요?
아래와 같은 방식으로도 해볼 수 있을 것 같아요.

```kotlin
val Saver: Saver<NewCardUiStateHolder, NewCardUiState> = 
    Saver(
        save = { .. },
        restore = { .. },
    )
```

chanho0908 replied:
[aa0bf2a](https://github.com/woowacourse/android-payments/pull/68/commits/aa0bf2a0ce427e523925b0f2a619b0c7a07a5f96)

NewCardUiState를 직렬화 가능하도록 만들면 오이가 말씀해주신 방법처럼 사용하는 방법을 몰랐네요
좋은 의건 감사합니다 🥒  👍

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardScreen.kt:
```diff
@@ -112,39 +137,61 @@ fun NewCardScreen(
+                    .padding(top = 18.dp),
         )
+
+        if (uiState.company is CardCompanyState.Empty) {
+            BankSelectBottomSheet(
+                resourceProvider = resourceProvider,
+                onFinish = onFinishRequest,
+                onBankSelect = { bankType ->
+                    onCardChange(NewCardUiEvent.OnChangeBankType(bankType))
+                },
+            )
+        }
```
카드사 선택 여부를 `sealed interface`를 이용해 상태로 나타낸 것은 좋은 아이디어인 것 같아요!

다만, 카드사 선택 (또는 바텀 시트가 열렸는지) 에 대한 여부를 `Boolean` 타입으로 나타냈다면, 
여러 곳에서 코드를 더 간결하게 작성할 수 있을 것 같다는 생각이 들어요.
상태가 다양하지 않고 예 혹은 아니오만 존재하기 때문에 이렇게 느낀 것 같아요.

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsScreen.kt:
```diff
@@ -124,23 +130,40 @@ fun CardsScreen(
+            CardsUiState.EMPTY ->
+                EmptyCardContent(resourceProvider, onClickCard)
+
+            is CardsUiState.SINGLE ->
+                SingleCardComponent(
+                    resourceProvider,
+                    CardState.Registered(uiState.state),
+                    onClickCard,
+                )
+
+            is CardsUiState.MULTIPLE ->
+                MultipleCardContent(
+                    resourceProvider,
+                    uiState.state.map { CardState.Registered(it) },
+                    onClickCard,
+                )
```
`SingleCardComponent`, `MultipleCardContent`
페토가 생각하는 Component와 Content의 차이는 무엇인가요?

chanho0908 replied:
[f4d9b2e](https://github.com/woowacourse/android-payments/pull/68/commits/f4d9b2e5195b6cf4b570654b34761cec292995c9)

Content는 다른 컴포저블에 후행 람다로 컴포저블을 전달할 때 사용되는 컴포즈의 네이밍 규약이라고 생각해요

특별한 의도를 둔건 아니고, 네이밍을 할 때 생각이 없이 지은 것 같네요 😓

꼼꼼하게 봐주셔서 감사합니다 🥒

cucumber99 reviewed (changes requested):
안녕하세요 페토~
역시 3단계도 깔끔하게 구현해주셨네요!
페토의 의견이 궁금해서 남긴 리뷰도 몇 있으니 꼭 답해주시고,
의논하고 싶은 내용이 있으시다면 DM 주시거나 랜딩으로 찾아오세요~~

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardUiStateHolder.kt:
```diff
@@ -26,42 +26,39 @@ class CardUiStateHolder(
+
         val Saver: Saver<CardUiStateHolder, Any> =
-            listSaver(
+            mapSaver(
                 save = { holder ->
-                    when (val uiState = holder.uiState) {
```
https://github.com/woowacourse/android-payments/pull/68#discussion_r2357629199

이 리뷰 참고해주세요!

chanho0908 committed:
refactor: `CompanyResourceProvider`를 `CardCompany` 확장 함수로 변경

chanho0908 committed:
refactor: NewCardUiState를 Parcelable로 변경하고 Saver 로직 간소화

chanho0908 committed:
refactor: `MultipleCardContent`를 `MultipleCardComponent`로 이름 변경

chanho0908 committed:
refactor: 카드사 선택 BottomSheet 로직 변경

- 카드사 선택 상태를 NewCardScreen에서 관리하도록 변경
- 카드사 미선택 시 BottomSheet가 표시되도록 수정
- OnChangeBankType을 OnChangeCardCompany로 이벤트명 변경 및 관련 로직 수정

chanho0908 committed:
test: 테스트 케이스 오류로 통과되지 않는 테스트 수정

chanho0908 committed:
test: 프로덕션 변경사항이 적용되지 않은 테스트 코드 수정

chanho0908 committed:
refactor: 카드사 이름이 없는 경우 카드 이름 컴포저블이 렌더링 되지 않게 수정

chanho0908 committed:
refactor: `PaymentCard`의 `onClick` 파라미터 제거 및 `clickable` Modifier 호출 위치 변경

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -1,36 +1,30 @@
-                        resourceProvider.getSignatureColor(
-                            card.card.company,
-                        )
+                Alignment.CenterStart to card.card.company.toSignatureColor()
         }
```
```kotlin
card.card.company.toSignatureColor()
```
🤔

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardScreen.kt:
```diff
@@ -20,35 +24,69 @@ import woowacourse.payments.ui.component.CardNumberTextField
+            NewCardUiStateHolder()
+        }
+    val uiState = newCardUiStateHolder.uiState
+    val bottomSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
+
```
`skipPartiallyExpanded`를 true로 설정하면 어떤 동작을 하나요?
또, 화면 외부 영역을 클릭했을 때 dismiss가 되게끔 의도한 이유는 무엇인가요?

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardScreen.kt:
```diff
@@ -20,35 +24,69 @@ import woowacourse.payments.ui.component.CardNumberTextField
         )
     }
+
+    BankSelectBottomSheet(
+        modalBottomSheetState = bottomSheetState,
+        onFinish = onFinishRequest,
+        onCardCompanySelect = { company ->
+            newCardUiStateHolder.updateCard(
+                NewCardUiEvent.OnChangeCardCompany(CardCompanyState.Selected(company)),
+            )
+        },
+    )
```
외부 영역을 클릭하거나 뒤로 가기 버튼을 눌렀을 때 `finish()`가 호출되는군요!
이렇게 동작하게끔 코드를 작성한 이유가 궁금해요~

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/BankSelectBottomSheet.kt:
```diff
@@ -0,0 +1,75 @@
+        verticalArrangement = Arrangement.spacedBy(23.dp),
+        maxItemsInEachRow = COLUMN_COUNT,
+    ) {
+        CardCompany.entries.forEachIndexed { index, company ->
+            CardCompanyIcon(
+                company = company,
+                onClick = onClick,
+            )
+        }
```
`index`를 사용하고 있지 않은데, 그냥 `forEach`를 사용해도 되지 않을까요?

cucumber99 reviewed (approved):
리팩터링 고생 많으셨어요~
리뷰 확인해주시고 다음 단계에서 반영해주세요!
